solana/indexer/backfill.go

```diff
@@ -0,0 +1,246 @@
+	var lastIndexedSig solana.Signature
+	foundIntersection := false
+	before := txRange.before
+
+	logger := s.logger.With(
+		zap.String("indexerSource", "rpcBackfill"),
+		zap.String("address", address.String()),
+	)
+	logger.Info("starting backfill for address",
+		zap.String("before", txRange.before.String()),
+		zap.String("until", txRange.until.String()),
+		zap.Uint64("fromSlot", fromSlot),
+		zap.Uint64("toSlot", toSlot),
+	)
+
+	limit := 1000
+	opts := rpc.GetSignaturesForAddressOpts{
+		Commitment:     rpc.CommitmentConfirmed,
+		Until:          txRange.until,
+		Limit:          &limit,
+		MinContextSlot: &toSlot,
+	}
+
+	for !foundIntersection {
+		select {
+		case <-ctx.Done():
+			logger.Error("failed to find intersection", zap.Error(ctx.Err()))
+			return
+		default:
+		}
+
+		opts.Before = before
+		res, err := withRetries(func() ([]*rpc.TransactionSignature, error) {
+			return s.rpcClient.GetSignaturesForAddressWithOpts(ctx, address, &opts)
+		}, 5, time.Second*1)
+		if err != nil {
+			logger.Error("failed to get signatures for address", zap.Error(err))
+			continue
+		}
+
+		if len(res) == 0 {
+			logger.Info("no transactions left to index.")
+			break
+		}
+
+		for _, sig := range res {
+			logger := logger.With(zap.String("signature", sig.Signature.String()))
+
+			select {
+			case <-ctx.Done():
+				logger.Error("failed to process signature", zap.Error(ctx.Err()))
+				return
+			default:
+			}
+
+			if sig.Slot < fromSlot {
+				foundIntersection = true
+				logger.Info("found intersection with transaction range")
+				break
+			}
+
+			// Skip error transactions
+			if sig.Err != nil {
+				lastIndexedSig = sig.Signature
+				continue
+			}
+
+			// Skip zero signatures
+			if sig.Signature.IsZero() {
+				continue
+			}
+
+			// Skip existing transactions
+			var exists bool
+			checkSql := `SELECT EXISTS(SELECT 1 FROM sol_token_account_balance_changes WHERE signature = @signature)`
+			err := s.pool.QueryRow(context.Background(), checkSql, pgx.NamedArgs{
+				"signature": sig.Signature,
+			}).Scan(&exists)
+
+			if err != nil {
+				logger.Error("failed to check if signature exists", zap.Error(err))
+				continue
+			}
+			if exists {
+				lastIndexedSig = sig.Signature
+				continue
+			}
+
+			err = s.ProcessSignature(ctx, sig.Slot, sig.Signature, logger)
+			if err != nil {
+				logger.Error("failed to process signature", zap.Error(err))
+			}
+
+			lastIndexedSig = sig.Signature
+
+			// sleep for a bit to avoid hitting rate limits
+			time.Sleep(time.Millisecond * time.Duration(TRANSACTION_DELAY_MS))
+		}
+
+		before = lastIndexedSig
+		logger.Info("finished transaction batch",
+			zap.Int("count", len(res)),
+		)
+
+		// sleep for a bit to avoid hitting rate limits
+		time.Sleep(time.Millisecond * time.Duration(BATCH_DELAY_MS))
+	}
+	insertBackfillCheckpoint(ctx, s.pool, fromSlot, toSlot, address.String())
```

Comment:
I noticed this function can return an error but we aren't doing anything with it here?
Just a reminder to do a pass and make sure we're surfacing errors if it's appropriate.

Reply:
great catch